### PR TITLE
mds: defer messages to bootstrapping ranks

### PR DESCRIPTION
**Pull request by batrick**

### CODE REVIEW

**Comment by batrick** [comment]
jenkins test api

**Comment by batrick** [comment]
https://tracker.ceph.com/projects/cephfs/wiki/Master#2021-November-9

File: qa/suites/fs/thrash/workloads/overrides/races.yaml [hunk 1]
```diff
@@ -0,0 +1,5 @@
+overrides:
+  ceph:
+    conf:
+      mds:
+        mds_sleep_rank_change: 5.0
```
Comment (vshankar) [inline qa/suites/fs/thrash/workloads/overrides/races.yaml:5]: Hmmm... 5 microseconds is most likely not going to simulate any race IMO (the test does a config_set for 10 seconds).

Should this be at-least 1 second to catch races? (in which case we can just use sleep()...)

File: src/mds/MDSRank.cc [hunk 2]
```diff
@@ -2437,6 +2462,33 @@ void MDSRankDispatcher::handle_mds_map(
...
+      waiting_for_bootstrapping_peer.erase(r);
+    }
+  }
+  /* for testing... */
+  if (unlikely(g_conf().get_val<bool>("mds_connect_bootstrapping"))) {
+    std::set<mds_rank_t> bootstrapping;
+    mdsmap->get_mds_set(bootstrapping, MDSMap::STATE_REPLAY);
+    mdsmap->get_mds_set(bootstrapping, MDSMap::STATE_CREATING);
+    mdsmap->get_mds_set(bootstrapping, MDSMap::STATE_STARTING);
+    for (const auto& r : bootstrapping) {
```
Comment (vshankar) [inline src/mds/MDSRank.cc:2486]: nit: const auto& rank?

File: src/mds/MDSRank.h [hunk 3]
```diff
@@ -583,8 +586,9 @@ class MDSRank {
     MDSContext::que replay_queue;
     bool replaying_requests_done = false;
 
-    std::map<mds_rank_t, MDSContext::vec > waiting_for_active_peer;
-    std::map<epoch_t, MDSContext::vec > waiting_for_mdsmap;
+    std::map<mds_rank_t, MDSContext::vec> waiting_for_active_peer;
+    std::map<mds_rank_t, MDSContext::vec> waiting_for_bootstrapping_peer;
```
Comment (vshankar) [inline src/mds/MDSRank.h:590]: nit: waiting_for_bootstrapped_peer?

**Review by vshankar — COMMENTED** [review]
mostly minor nits.

File: src/mds/MDSRank.cc [hunk 4]
```diff
@@ -2247,6 +2263,15 @@ void MDSRankDispatcher::handle_mds_map(
 
   if (oldstate != state) {
     // update messenger.
+    auto sleep_rank_change = g_conf().get_val<double>("mds_sleep_rank_change");
+    if (unlikely(sleep_rank_change > 0)) {
+      /* This is to trigger a race where another rank tries to connect to this
```
[the remark this answers is not in the document]
Comment (batrick) [reply src/mds/MDSRank.cc:2268]: fixed

File: src/mds/MDSRank.cc [hunk 5]
```diff
@@ -2437,6 +2462,33 @@ void MDSRankDispatcher::handle_mds_map(
...
+      if (state > MDSMap::STATE_REPLAY) {
+        queue_waiters(queue);
+        erase.push_back(rank);
+      }
+    }
+    for (const auto& r : erase) {
+      waiting_for_bootstrapping_peer.erase(r);
+    }
+  }
+  /* for testing... */
```
[the remark this answers is not in the document]
Comment (batrick) [reply src/mds/MDSRank.cc]: fixed

File: qa/suites/fs/thrash/workloads/overrides/races.yaml (hunk 1 above)
Comment (batrick) [reply qa/suites/fs/thrash/workloads/overrides/races.yaml]: Ya, that's an oops. Should be 5e6. Thanks.

File: src/mds/MDSRank.cc (hunk 2 above)
Comment (batrick) [reply src/mds/MDSRank.cc]: sure, fixed

File: src/mds/MDSRank.h (hunk 3 above)
Comment (batrick) [reply src/mds/MDSRank.h:590]: I actually deliberately chose "bootstrapping" because that's what the peer is doing. It could go either way though...

Comment (vshankar) [reply src/mds/MDSRank.h:590]: ack

**Review by vshankar — APPROVED** [review]

**Comment by vshankar** [comment]
@batrick I'll include this in my testing branch.

**Comment by batrick** [comment]
> @batrick I'll include this in my testing branch.

Venky, I took a glance at this PR again and noticed vim had wrecked the whitespace on some code while formatting a comment. I fixed the formatting and did a rebase.

**Comment by vshankar** [comment]
> Venky, I took a glance at this PR again and noticed vim had wrecked the whitespace on some code while formatting a comment. I fixed the formatting and did a rebase.

That's fine, Patrick.

**Comment by batrick** [comment]
jenkins test make check

**Comment by vshankar** [comment]
Failures are - https://tracker.ceph.com/projects/cephfs/wiki/Master#2021-Nov-30

**Pull request merged**